test/unit/discover-providers-test.js

```diff
@@ -125,5 +222,29 @@ test('discoverProvider createFromOpts', function t(assert) {
         var discoverProvider = discoverProviderFactory.createFromOpts();
         assert.equal(discoverProvider, null);
         assert.end();
-    })
+    });
+
+    assert.test('opts with retry - returns retrying discover provider', function t(assert) {
+        assert.plan(5);
+
+        var first = true;
+        var fn = function exampleDiscoverProvider(cb) {
+            if(first) {
+                first = false;
+                assert.pass('inner provider called once');
+                cb(null, null);
```

Comment:
Maybe test with an empty list as well. I think it is more realistic to expect an empty list then `null` if there are no peers.